File: wordPattern.java

import java.util.HashMap;

public class wordPattern {
    // 290. Word Pattern
    public boolean wordPattern(String pattern, String s) {
        HashMap map_index = new HashMap();
        String[] words = s.split(" ");
        if (words.length != pattern.length()) return false;

        for (Integer i = 0; i < pattern.length(); i++) {
            String word = words[i];
            char c = pattern.charAt(i);
            if (!map_index.containsKey(word)) {
                map_index.put(word, i);
            }
            if (!map_index.containsKey(c)) {
                map_index.put(c, i);
            }

            if (map_index.get(word) != map_index.get(c)) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        
    }
}

File: wordPattern.py
class Solution:
    def __init__(self) -> None:
        pass

    def wordPattern(self, pattern: str, s: str) -> bool:
        map = {}
        s = s.split(' ')
        if len(s) != len(pattern):
            return False

        for i, (c, w) in enumerate(zip(s, pattern)):
            c = "char_"+c
            w = "word_"+w
            if c not in map:
                map[c] = i
            if w not in map:
                map[w] = i
            if map[c] != map[w]:
                return False
        
        return True

    
def main():
    solution = Solution()
    print(solution.wordPattern("aaaa", "dog cat cat dog"))

if __name__ == "__main__":
    main()